### PR TITLE
Auth/PM-5501 - VaultTimeoutSettingsService State Provider Migration

### PR DESCRIPTION
**Pull request by JaredSnider-Bitwarden**
## Type of change

```
- [ ] Bug fix
- [X] New feature development
- [X] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Resolve [PM-5501 ](https://bitwarden.atlassian.net/browse/PM-5501) by migrating the vault timeout settings out of the `StateService` into `StateProvider` with the `VaultTimeoutSettingsService` owning that state. 

### Core `VaultTimeoutSettingsService` changes

- **libs/common/src/platform/state/state-definitions.ts:** Add new `VAULT_TIMEOUT_SETTINGS_DISK_LOCAL` state definition
- **libs/common/src/services/vault-timeout/vault-timeout-settings.state.ts:** Add `UserKeyDefinitions` for vault timeout action and vault timeout. 
- **libs/common/src/services/vault-timeout/vault-timeout-settings.state.spec.ts:** Test deserialization of `UserKeyDefinitions`
- **libs/common/src/services/vault-timeout/vault-timeout-settings.service.ts:**
    - Wire up `StateProvider` while significantly refactoring the retrieval logic for both the vault timeout action and vault timeout to be truly reactive and more single responsibility. 
    - Also, moved towards a design which requires a user id for all sets and gets to be safer.
- **libs/common/src/services/vault-timeout/vault-timeout-settings.service.spec.ts:** Update tests
- **libs/common/src/abstractions/vault-timeout/vault-timeout-settings.service.ts:** Update abstraction

### Dependency Updates
- **apps/cli/src/bw.ts:** Move `vaultTimeoutSettingsService` instantiation + deps above new `ApiService` usage. Add `VaultTimeoutSettingsService` to `LoginStrategyService`
- **apps/browser/src/background/main.background.ts:** Move `vaultTimeoutSettingsService` instantiation + deps above new `ApiService` usage. Add `VaultTimeoutSettingsService` to `LoginStrategyService` and `IdleBackground`
- **apps/browser/src/background/service-factories/vault-timeout-settings-service.factory.ts:** Add new deps
- **apps/browser/src/auth/background/service-factories/login-strategy-service.factory.ts:** Add `VaultTimeoutSettingsService` to `LoginStrategyService`
- **apps/browser/src/platform/background/service-factories/api-service.factory.ts:** Add `vaultTimeoutSettingsServiceFactory` to `ApiServiceFactory` by replacing `StateServiceFactory`
- **apps/cli/src/platform/services/node-api.service.ts:** Replace `StateService` dep with `VaultTimeoutSettingsService`
- **libs/angular/src/services/jslib-services.module.ts:** Update `VaultTimeoutSettingsService`, `ApiService`, and `LoginStrategyService` deps

### `StateService` cleanup

- **libs/common/src/platform/abstractions/state.service.ts:** Remove vault timeout data setters and getters
- **libs/common/src/platform/services/state.service.ts:** Remove vault timeout data setters and getters
- **libs/common/src/platform/models/domain/account.ts:** Remove vault timeout data from account
- **libs/common/src/platform/models/domain/global-state.ts:** Remove vault timeout data from global state

### User Data Migration

- **libs/common/src/state-migrations/migrations/59-migrate-vault-timeout-settings-svc-to-state-provider.ts:** Migrate user vault timeout data into `StateProvider`. Note: we do not persist global vault timeout data as it was already almost fully deprecated.
- **libs/common/src/state-migrations/migrations/59-migrate-vault-timeout-settings-svc-to-state-provider.spec.ts:** Test migration
- **libs/common/src/state-migrations/migrate.ts:** Register migration

### Reference Updates (replacing `StateService` methods with calls to `VaultTimeoutSettingsService`)
- **apps/browser/src/background/idle.background.ts**
- **apps/browser/src/popup/settings/settings.component.ts** 
- **apps/desktop/src/app/app.component.ts**
- **apps/web/src/app/settings/preferences.component.ts**
- **libs/auth/src/common/login-strategies/login.strategy.ts:** (requires updates to all login strategies)
- **libs/auth/src/common/login-strategies/auth-request-login.strategy.spec.ts**
- **libs/auth/src/common/login-strategies/login.strategy.spec.ts**
- **libs/auth/src/common/login-strategies/password-login.strategy.spec.ts**
- **libs/auth/src/common/login-strategies/password-login.strategy.ts**
- **libs/auth/src/common/login-strategies/sso-login.strategy.spec.ts**
- **libs/auth/src/common/login-strategies/sso-login.strategy.ts**
- **libs/auth/src/common/login-strategies/user-api-login.strategy.spec.ts**
- **libs/auth/src/common/login-strategies/user-api-login.strategy.ts**
- **libs/auth/src/common/login-strategies/webauthn-login.strategy.spec.ts**
- **libs/auth/src/common/login-strategies/webauthn-login.strategy.ts**
- **libs/auth/src/common/services/login-strategies/login-strategy.service.spec.ts**
- **libs/auth/src/common/services/login-strategies/login-strategy.service.ts**
- **libs/common/src/platform/services/crypto.service.spec.ts**
- **libs/common/src/platform/services/crypto.service.ts**
- **libs/common/src/platform/services/system.service.ts**
- **libs/common/src/services/api.service.ts**
- **libs/common/src/services/vault-timeout/vault-timeout.service.ts**
- **libs/common/src/services/vault-timeout/vault-timeout.service.spec.ts** 

## Screenshots

Note: I tested both vault timeout actions with 1 min timers on all 3 clients which support vault timeout settings (web, desktop, and browser), but didn't record as it would be 1 min of sitting per client per action. 

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

### CODE REVIEW

**Review by jlf0dev — COMMENTED** [review]
Great job! Just had some comments on the service itself but everything else looks good

File: libs/common/src/services/vault-timeout/vault-timeout-settings.service.ts [hunk 1]
```diff
@@ -88,72 +107,134 @@ export class VaultTimeoutSettingsService implements VaultTimeoutSettingsServiceA
...
+        return from(
+          this.determineVaultTimeoutAction(
+            userId,
+            currentVaultTimeoutAction,
+            maxVaultTimeoutPolicy,
+          ),
+        ).pipe(
+          tap((vaultTimeoutAction: VaultTimeoutAction) => {
+            // As a side effect, set the new value in the state if it's different from the current
+            // We want to avoid having a null timeout action always so we set it to the default if it is null
```
[the remark this answers is not in the document]
Comment (JaredSnider-Bitwarden) [reply libs/common/src/services/vault-timeout/vault-timeout-settings.service.ts:201]: I clarified that the tap is fed by the determine methods in both get methods.

File: libs/common/src/services/vault-timeout/vault-timeout-settings.service.ts [hunk 2]
```diff
@@ -88,72 +107,134 @@ export class VaultTimeoutSettingsService implements VaultTimeoutSettingsServiceA
...
-      }
+      // Remove negative values, and ensure it's smaller than maximum allowed value according to policy
+      let policyCompliantTimeout = Math.min(currentVaultTimeout, maxVaultTimeoutPolicyData.minutes);
 
-      // TODO @jlf0dev: Can we move this somwhere else? Maybe add it to the initialization process?
-      // ( Apparently I'm the one that reviewed the original PR that added this :) )
-      // We really shouldn't need to set the value here, but multiple services relies on this value being correct.
-      if (vaultTimeout !== timeout) {
-        await this.stateService.setVaultTimeout(timeout, { userId });
+      if (currentVaultTimeout == null || policyCompliantTimeout < 0) {
```
[the remark this answers is not in the document]
Comment (JaredSnider-Bitwarden) [reply libs/common/src/services/vault-timeout/vault-timeout-settings.service.ts]: Good call! Done!

File: libs/common/src/services/vault-timeout/vault-timeout-settings.service.ts [hunk 3]
```diff
@@ -88,72 +107,134 @@ export class VaultTimeoutSettingsService implements VaultTimeoutSettingsServiceA
...
 
-    if (policies?.length) {
-      // Remove negative values, and ensure it's smaller than maximum allowed value according to policy
-      let timeout = Math.min(vaultTimeout, policies[0].data.minutes);
+  private async determineVaultTimeout(
+    userId: UserId,
+    currentVaultTimeout: number,
+    maxVaultTimeoutPolicy: Policy | null,
+  ): Promise<number> {
+    if (maxVaultTimeoutPolicy) {
```
[the remark this answers is not in the document]
Comment (JaredSnider-Bitwarden) [reply libs/common/src/services/vault-timeout/vault-timeout-settings.service.ts]: Good call! Done!

File: libs/common/src/services/vault-timeout/vault-timeout-settings.service.ts [hunk 4]
```diff
@@ -27,30 +45,31 @@ export class VaultTimeoutSettingsService implements VaultTimeoutSettingsServiceA
...
+    private stateProvider: StateProvider,
+    private logService: LogService,
   ) {}
 
-  async setVaultTimeoutOptions(timeout: number, action: VaultTimeoutAction): Promise<void> {
+  async setVaultTimeoutOptions(
+    userId: UserId,
+    timeout: number,
+    action: VaultTimeoutAction,
+  ): Promise<void> {
```
[the remark this answers is not in the document]
Comment (JaredSnider-Bitwarden) [reply libs/common/src/services/vault-timeout/vault-timeout-settings.service.ts]: Good call! Done!

[the remark this answers is not in the document]
Comment (JaredSnider-Bitwarden) [reply libs/common/src/services/vault-timeout/vault-timeout-settings.service.ts]: Thank you for catching the missing `userId` validation!

**Review by jlf0dev — APPROVED** [review]
👏

**Comment by JaredSnider-Bitwarden** [comment]
Additional context for reviewers as to why we had to make more changes:  
- @justindbaur discovered that we had client specific default values on account initialization for the vault timeout. I had to rework the vault timeout retrieval to respect the client specific default.  While doing so and while testing the changes, I realized that I needed to rework the default logic for the vault timeout action to ensure the previous default of lock was respected.

File: apps/browser/src/popup/settings/settings.component.ts [hunk 5]
```diff
@@ -132,20 +135,24 @@ export class SettingsComponent implements OnInit {
       { name: this.i18nService.t("thirtyMinutes"), value: 30 },
       { name: this.i18nService.t("oneHour"), value: 60 },
       { name: this.i18nService.t("fourHours"), value: 240 },
-      // { name: i18nService.t('onIdle'), value: -4 },
-      // { name: i18nService.t('onSleep'), value: -3 },
+      // { name: i18nService.t('onIdle'), value: "onIdle" },
+      // { name: i18nService.t('onSleep'), value: "onSleep" },
```
Comment (cagonzalezcs) [inline apps/browser/src/popup/settings/settings.component.ts:139]: Nit - Should this be removed?

File: libs/angular/src/components/settings/vault-timeout-input.component.ts [hunk 6]
```diff
@@ -198,12 +199,24 @@ export class VaultTimeoutInputComponent
...
+        return true;
+      }
+
+      if (typeof vaultTimeoutOption.value === "number") {
+        // Include numeric values that are less than or equal to the policy minutes
+        return vaultTimeoutOption.value <= this.vaultTimeoutPolicy.data.minutes;
+      } else {
+        // Exclude all string cases when there's a numeric policy defined
+        return false;
+      }
```
Comment (cagonzalezcs) [inline libs/angular/src/components/settings/vault-timeout-input.component.ts:214]:
```suggestion
      }
        
      // Exclude all string cases when there's a numeric policy defined
      return false;
```

Nit - you don't really need this `else` branch. The early returns above this statement make it irrelevant

File: apps/browser/src/popup/settings/settings.component.ts (hunk 5 above)
Comment (JaredSnider-Bitwarden) [reply apps/browser/src/popup/settings/settings.component.ts:139]: Cleaned up the commented out options on browser and desktop!

File: apps/browser/src/background/main.background.ts [hunk 7]
```diff
@@ -556,12 +556,29 @@ export default class MainBackground {
...
+    this.vaultTimeoutSettingsService = new VaultTimeoutSettingsService(
+      this.userDecryptionOptionsService,
+      this.cryptoService,
+      this.tokenService,
+      this.policyService,
+      this.stateService,
+      this.biometricStateService,
+      this.stateProvider,
+      this.logService,
+      "onRestart", // default vault timeout
```
[the remark this answers is not in the document]
Comment (JaredSnider-Bitwarden) [reply apps/browser/src/background/main.background.ts]: Thank you for showing me the way to handle this! Updated everything to reference the new const!

**Review by cagonzalezcs — APPROVED** [review]

**Comment by JaredSnider-Bitwarden** [comment]
Further notes for reviewers, we had more work to do because we discovered that the CLI wasn't setting a vault timeout so we had to enable migrations to have access to the client type they were executing on so we could successfully migrate the CLI's undefined vault timeouts to the new `"never"` value. I tested logging in on `main` in the CLI, and swapping over to this branch, and I ensured that the migration was seamless with the user's vault still being unlocked properly.

File: apps/browser/src/background/main.background.ts [hunk 8]
```diff
@@ -559,12 +560,29 @@ export default class MainBackground {
...
+    this.vaultTimeoutSettingsService = new VaultTimeoutSettingsService(
+      this.userDecryptionOptionsService,
+      this.cryptoService,
+      this.tokenService,
+      this.policyService,
+      this.stateService,
+      this.biometricStateService,
+      this.stateProvider,
+      this.logService,
+      VaultTimeoutStringType.OnRestart, // default vault timeout
```
Comment (cagonzalezcs) [inline apps/browser/src/background/main.background.ts:577]: 🎉

File: libs/common/src/types/vault-timeout.type.ts [hunk 9]
```diff
@@ -0,0 +1,17 @@
+// Note: the below comments are just for documenting what they used to be.
+export const VaultTimeoutStringType = {
```
Comment (cagonzalezcs) [inline libs/common/src/types/vault-timeout.type.ts:2]:
```suggestion
export const VaultTimeoutType = {
```
Maybe there's some distinct reason we're adding the explicitly state "String" data type here... but wouldn't it normally be extraneous information that is unnecessary to be concerned about?

File: apps/web/src/app/core/core.module.ts [hunk 10]
```diff
@@ -69,6 +71,12 @@ const safeProviders: SafeProvider[] = [
   safeProvider(RouterService),
   safeProvider(EventService),
   safeProvider(PolicyListService),
+  safeProvider({
+    provide: DEFAULT_VAULT_TIMEOUT,
+    deps: [PlatformUtilsServiceAbstraction],
+    useFactory: (platformUtilsService: PlatformUtilsServiceAbstraction): VaultTimeout =>
+      platformUtilsService.isDev() ? VaultTimeoutStringType.Never : 15,
```
Comment (cagonzalezcs) [inline apps/web/src/app/core/core.module.ts:78]: Should this hardcoded `15` value also reference an enum?

**Review by cagonzalezcs — APPROVED** [review]
A coupe of small nits, but not blocking for them. Nice work on this.

File: apps/web/src/app/core/core.module.ts (hunk 10 above)
Comment (JaredSnider-Bitwarden) [reply apps/web/src/app/core/core.module.ts:78]: I think you are correct that it and the settings / preferences components across our clients should reference a numeric enum for all the default timeouts that we show to the user. However, that's continuing to expand the scope of the PR into more tech debt than I want to tackle right now. I have created a tech debt ticket for this [PM-7841](https://bitwarden.atlassian.net/browse/PM-7841) for us to tackle later! Thank you for pointing it out!

File: apps/web/src/app/core/state/account.ts
Comment (MGibson1) [inline apps/web/src/app/core/state/account.ts:1]: 🥳

File: libs/auth/src/common/login-strategies/auth-request-login.strategy.spec.ts [hunk 11]
```diff
@@ -106,11 +111,27 @@ describe("AuthRequestLoginStrategy", () => {
...
+
+    const mockVaultTimeoutAction = VaultTimeoutAction.Lock;
+    const mockVaultTimeoutActionBSub = new BehaviorSubject<VaultTimeoutAction>(
+      mockVaultTimeoutAction,
+    );
+    vaultTimeoutSettingsService.getVaultTimeoutActionByUserId$.mockReturnValue(
+      mockVaultTimeoutActionBSub.asObservable(),
+    );
+
+    const mockVaultTimeout = 1000;
```
Comment (MGibson1) [inline libs/auth/src/common/login-strategies/auth-request-login.strategy.spec.ts:129]: This doesn't cause 1 second delays, does it?

File: libs/common/src/types/vault-timeout.type.ts
Comment (MGibson1) [inline libs/common/src/types/vault-timeout.type.ts:1]: 😍

File: libs/auth/src/common/login-strategies/login.strategy.ts [hunk 12]
```diff
@@ -201,10 +190,27 @@ export abstract class LoginStrategy {
...
     await this.userDecryptionOptionsService.setUserDecryptionOptions(
       UserDecryptionOptions.fromResponse(tokenResponse),
     );
 
+    const vaultTimeoutAction = await firstValueFrom(
+      this.vaultTimeoutSettingsService.getVaultTimeoutActionByUserId$(userId),
+    );
+    const vaultTimeout = await firstValueFrom(
+      this.vaultTimeoutSettingsService.getVaultTimeoutByUserId$(userId),
+    );
```
Comment (MGibson1) [inline libs/auth/src/common/login-strategies/login.strategy.ts:204]: How are these populated with default values on first login?

File: libs/common/src/state-migrations/migrations/61-migrate-vault-timeout-settings-svc-to-state-provider.ts [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+  onIdle: -4,
+};
+
+export enum ClientType {
+  Web = "web",
+  Browser = "browser",
+  Desktop = "desktop",
+  // Mobile = "mobile",
+  Cli = "cli",
+  // DirectoryConnector = "connector",
```
Comment (MGibson1) [inline libs/common/src/state-migrations/migrations/61-migrate-vault-timeout-settings-svc-to-state-provider.ts:64]: I think these commented values are here to indicate they exist, too, but an explanation is warranted.

File: libs/common/src/state-migrations/migrations/61-migrate-vault-timeout-settings-svc-to-state-provider.ts [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+      // into the new state provider framework.  It was originally a fallback but hasn't been used for years
+      // so this migration will clean up the global properties fully.
+
+      if (updatedAccount) {
+        // Save the migrated account only if it was updated
+        await helper.set(userId, account);
+      }
+    }
+
+    // await Promise.all([...accounts.map(({ userId, account }) => migrateAccount(userId, account))]);
```
Comment (MGibson1) [inline libs/common/src/state-migrations/migrations/61-migrate-vault-timeout-settings-svc-to-state-provider.ts:125]: Remove commented code and comment below explaining it

**Review by MGibson1 — CHANGES_REQUESTED** [review]
Very small clarity changes requested for now. I'd like to spend some more time on understanding the new vault-timeout service, but this is looking like a _huge_ cleanup of that area. thank you

File: libs/auth/src/common/login-strategies/auth-request-login.strategy.spec.ts (hunk 11 above)
Comment (JaredSnider-Bitwarden) [reply libs/auth/src/common/login-strategies/auth-request-login.strategy.spec.ts:129]: No, it's just a randomly picked numeric value to be returned for the sake of the tests. Further, the vault timeout value is always in minutes but the test suite runs near instantly.

File: libs/auth/src/common/login-strategies/login.strategy.ts (hunk 12 above)
Comment (JaredSnider-Bitwarden) [reply libs/auth/src/common/login-strategies/login.strategy.ts:204]: Vault timeout is defaulted via the `DEFAULT_VAULT_TIMEOUT` injection token and then applied in the `VaultTimeoutSettingsService`'s `determineVaultTimeout(...)` method when the vault timeout is null from state.

Vault timeout action is defaulted via the `VaultTimeoutSettingsService`'s `determineVaultTimeoutAction(...)` method to `lock` if no policy applies and the state value is null.  It used to be set to lock via the account init and then modified in the previous implementation.

File: libs/common/src/state-migrations/migrations/61-migrate-vault-timeout-settings-svc-to-state-provider.ts (hunk 13 above)
Comment (JaredSnider-Bitwarden) [reply libs/common/src/state-migrations/migrations/61-migrate-vault-timeout-settings-svc-to-state-provider.ts:64]: Hmm... it's copy pasta from the real client type enum. I've removed them entirely.

File: libs/common/src/state-migrations/migrations/61-migrate-vault-timeout-settings-svc-to-state-provider.ts (hunk 14 above)
Comment (JaredSnider-Bitwarden) [reply libs/common/src/state-migrations/migrations/61-migrate-vault-timeout-settings-svc-to-state-provider.ts:125]: That was an accidental check in of a change I made to make it easier to debug the migration. Thank you for catching that. I've reverted it back to use `Promise.all`!

File: libs/common/src/types/vault-timeout.type.ts (hunk 9 above)
Comment (cagonzalezcs) [reply libs/common/src/types/vault-timeout.type.ts:2]: Hm, I think given the circumstances behind switching from one data type to another, it makes sense to be more explicit here. 

I guess the ideal would be to eventually refactor out the `number` data type, right? Regardless, I think keeping the current structure in place makes sense.

**Review by cagonzalezcs — DISMISSED** [review]

File: libs/common/src/services/vault-timeout/vault-timeout-settings.service.ts [hunk 15]
```diff
@@ -27,30 +46,36 @@ export class VaultTimeoutSettingsService implements VaultTimeoutSettingsServiceA
...
     private biometricStateService: BiometricStateService,
+    private stateProvider: StateProvider,
+    private logService: LogService,
+    private defaultVaultTimeout: VaultTimeout,
   ) {}
 
-  async setVaultTimeoutOptions(timeout: number, action: VaultTimeoutAction): Promise<void> {
+  async setVaultTimeoutOptions(
+    userId: UserId,
+    timeout: VaultTimeout | null,
```
Comment (jlf0dev) [inline libs/common/src/services/vault-timeout/vault-timeout-settings.service.ts:56]: Question: should this allow `null` still?

File: libs/common/src/services/vault-timeout/vault-timeout-settings.service.ts [hunk 16]
```diff
@@ -88,72 +113,156 @@ export class VaultTimeoutSettingsService implements VaultTimeoutSettingsServiceA
...
-    return vaultTimeout;
+    // For numeric vault timeouts, ensure they are smaller than maximum allowed value according to policy
+    const policyCompliantTimeout = Math.min(currentVaultTimeout, maxVaultTimeoutPolicyData.minutes);
+
+    return policyCompliantTimeout;
   }
 
-  vaultTimeoutAction$(userId?: UserId) {
-    return defer(() => this.getVaultTimeoutAction(userId));
+  private async setVaultTimeoutAction(userId: UserId, action: VaultTimeoutAction): Promise<void> {
```
Comment (jlf0dev) [inline libs/common/src/services/vault-timeout/vault-timeout-settings.service.ts:179]: Question: should we check action for `null` here and throw?

File: libs/common/src/services/vault-timeout/vault-timeout-settings.service.ts [hunk 17]
```diff
@@ -184,7 +293,8 @@ export class VaultTimeoutSettingsService implements VaultTimeoutSettingsServiceA
       if (decryptionOptions?.hasMasterPassword != undefined) {
         return decryptionOptions.hasMasterPassword;
       }
```
Comment (jlf0dev) [inline libs/common/src/services/vault-timeout/vault-timeout-settings.service.ts:295]: I know I wrote this, but can you change it to be:
```suggestion
      !!decryptionOptions.hasMasterPassword
```
The else you added below means that we'd return undefined.

File: libs/common/src/services/vault-timeout/vault-timeout-settings.state.ts [hunk 18]
```diff
@@ -0,0 +1,24 @@
+import { VaultTimeoutAction } from "../../enums/vault-timeout-action.enum";
+import { UserKeyDefinition, VAULT_TIMEOUT_SETTINGS_DISK_LOCAL } from "../../platform/state";
+import { VaultTimeout } from "../../types/vault-timeout.type";
+
+/**
+ * Settings use disk storage and local storage on web so settings can persist after logout.
+ */
```
Comment (jlf0dev) [inline libs/common/src/services/vault-timeout/vault-timeout-settings.state.ts:7]: nit: Can you add a little more on why it needs to persist after logout?

**Review by jlf0dev — APPROVED** [review]
Fantastic work, adding the new vault timeout type makes everything so much more readable. I had a couple questions and nits, but nothing blocking 👏

File: libs/common/src/services/vault-timeout/vault-timeout-settings.service.ts (hunk 15 above)
Comment (JaredSnider-Bitwarden) [reply libs/common/src/services/vault-timeout/vault-timeout-settings.service.ts:56]: Nope! Great catch! Ty!

File: libs/common/src/services/vault-timeout/vault-timeout-settings.service.ts (hunk 16 above)
Comment (JaredSnider-Bitwarden) [reply libs/common/src/services/vault-timeout/vault-timeout-settings.service.ts:191]: Fixed and tested! Great suggestion!

File: libs/common/src/services/vault-timeout/vault-timeout-settings.state.ts (hunk 18 above)
Comment (JaredSnider-Bitwarden) [reply libs/common/src/services/vault-timeout/vault-timeout-settings.state.ts:7]: Done!

File: libs/common/src/services/vault-timeout/vault-timeout-settings.service.ts (hunk 17 above)
Comment (JaredSnider-Bitwarden) [reply libs/common/src/services/vault-timeout/vault-timeout-settings.service.ts]: Ah, good catch. Done!

**Review by jlf0dev — APPROVED** [review]

**Review by MGibson1 — APPROVED** [review]

**Review by jprusik — APPROVED** [review]

File: libs/common/src/auth/services/token.service.ts [hunk 19]
```diff
@@ -160,14 +160,23 @@ export class TokenService implements TokenServiceAbstraction {
...
     clientIdClientSecret?: [string, string],
   ): Promise<void> {
     if (!accessToken) {
       throw new Error("Access token is required.");
     }
 
+    // Can't check for nullish b/c 0 is a valid value
+    if (vaultTimeout == null) {
+      throw new Error("Vault Timeout is required.");
+    }
```
Comment (jlf0dev) [inline libs/common/src/auth/services/token.service.ts:174]: nit:
```suggestion
    // Can't check for falsey b/c 0 is a valid value
    if (vaultTimeout == null) {
      throw new Error("Vault Timeout is required.");
    }
```

File: libs/common/src/auth/services/token.service.ts [hunk 20]
```diff
@@ -886,10 +895,19 @@ export class TokenService implements TokenServiceAbstraction {
...
-    vaultTimeout: VaultTimeout | null,
+    vaultTimeout: VaultTimeout,
     useSecureStorage: boolean,
   ): Promise<TokenStorageLocation> {
-    if (vaultTimeoutAction === VaultTimeoutAction.LogOut && vaultTimeout != null) {
+    if (vaultTimeout == null) {
+      throw new Error(
+        "TokenService - determineStorageLocation: We expect the vault timeout to always exist at this point.",
+      );
+    }
```
Comment (jlf0dev) [inline libs/common/src/auth/services/token.service.ts:905]: suggestion: would be nice to have a nullish check for `vaultTimeoutAction` as well.

File: libs/common/src/auth/services/token.service.ts [hunk 21]
```diff
@@ -272,7 +282,7 @@ export class TokenService implements TokenServiceAbstraction {
   private async _setAccessToken(
     accessToken: string,
     vaultTimeoutAction: VaultTimeoutAction,
-    vaultTimeout: number | null,
+    vaultTimeout: VaultTimeout | null,
```
Comment (jlf0dev) [inline libs/common/src/auth/services/token.service.ts:285]: Question: are all these `VaultTimeout | null` in the service and abstraction accurate? I thought we're not allowing that here anymore

File: libs/common/src/auth/services/token.service.ts (hunk 19 above)
Comment (JaredSnider-Bitwarden) [reply libs/common/src/auth/services/token.service.ts]: Fixed

File: libs/common/src/auth/services/token.service.ts (hunk 20 above)
Comment (JaredSnider-Bitwarden) [reply libs/common/src/auth/services/token.service.ts]: Fixed

File: libs/common/src/auth/services/token.service.ts (hunk 21 above)
Comment (JaredSnider-Bitwarden) [reply libs/common/src/auth/services/token.service.ts]: Thank you for catching that. I've removed them all.

**Review by jlf0dev — APPROVED** [review]

**Review by MGibson1 — APPROVED** [review]

**Review by jprusik — APPROVED** [review]

**Pull request merged**